extends Control
class_name DebugConsole

var command_history = []

@export var input_box: LineEdit
@export var history_box: RichTextLabel
# å¯¹è¯ç®¡ç†å™¨
@export var dialog_manager: DialogueManager
# å¯¹è¯æ¥å£
@export var dialog_interface: DialogueInterface
# éŸ³é¢‘ç®¡ç†å™¨
@export var audio_interface: DialogAudioInterface
# æ–‡ä»¶åŠ è½½å™¨
@onready var ks_file_loader: FileDialog = $LoadKSFileDialog
func _ready():
	input_box.text_submitted.connect(_on_input_text_entered)
	_update_output("[color=CYAN]æ§åˆ¶å°è¾“å…¥ï¼šhelp æŸ¥çœ‹å‘½ä»¤åˆ—è¡¨[/color]")
	
func _on_input_text_entered(text):
	# å¤„ç†ç”¨æˆ·è¾“å…¥
	var input_command = text.strip_edges() # å»é™¤é¦–å°¾ç©ºæ ¼
	_update_output("[color=CYAN]>> [/color]" + input_command) # å°†ç”¨æˆ·è¾“å…¥æ·»åŠ åˆ°è¾“å‡ºåŒºåŸŸ
	match input_command.split(" ")[0]:
		"next":
			_execute_next_command()
		"stop":
			_execute_stop_command()
		"start":
			_execute_start_command()
		"jump_cur":
			var id = input_command.split(" ")[1].to_int()
			print(id)
			_execute_jump_cur_command(id)
		"jump_data":
			var id = str(input_command.split(" ")[1])
			_execute_jump_data_command(id)
		"set_name":
			var name = String(input_command.split(" ")[1])
			_execute_set_name_command(name)
		"set_content":
			var content = String(input_command.split(" ")[1])
			_execute_set_content_command(content)
		"info":
			_execute_info_command()
		"help":
			_execute_help_command()
		"clean":
			_execute_clean_command()
		"history":
			_execute_history_command()
		"clean_history":
			_execute_clean_history_command()
		"list_dialog_datas":
			_execute_list_dialog_datas_command()
		"load_ks":
			_execute_load_ks_command()
		"":
			pass
		_:
			_update_output("[color=ORANGE]å‘½ä»¤æ²¡æœ‰æ‰¾åˆ°ğŸ« : [/color]" + input_command)

	command_history.append(input_command) # å°†å‘½ä»¤æ·»åŠ åˆ°å†å²è®°å½•

func _update_output(text):
	# æ¸…ç©ºè¾“å…¥å†…å®¹
	input_box.clear()
	# æ›´æ–°è¾“å‡ºåŒºåŸŸ
	history_box.text += text + "\n" # åœ¨æ–‡æœ¬æ˜¾ç¤ºåŒºåŸŸæœ«å°¾æ·»åŠ æ–°å†…å®¹
	history_box.scroll_to_line(history_box.get_line_count()) # æ»šåŠ¨åˆ°æœ€åä¸€è¡Œ

func _execute_next_command():
	# æ‰§è¡Œä¸‹ä¸€å¥æŒ‡ä»¤
	dialog_manager._continue()

func _execute_stop_command():
	# æ‰§è¡ŒstopæŒ‡ä»¤
	dialog_manager._stop_dialogue()
	_update_output("[color=SEASHELL]åœæ­¢å¯¹è¯...[/color]")

func _execute_start_command():
	# æ‰§è¡ŒstartæŒ‡ä»¤
	dialog_manager._start_dialogue()
	_update_output("[color=SEASHELL]å¼€å¯å¯¹è¯...[/color]")

func _execute_jump_cur_command(id):
	# æ‰§è¡Œjump_curæŒ‡ä»¤
	var error = dialog_manager.debug_jump_curline(id)
	if error:
		_update_output("[color=SNOW]è·³è½¬åˆ°ç¬¬" + str(id) + "å¥å¯¹è¯")
	else:
		_update_output("[color=ORANGE]æ²¡æœ‰è¿™å¥å¯¹è¯: [/color]")

func _execute_jump_data_command(id: String):
	# æ‰§è¡Œjump_dataæŒ‡ä»¤
	_update_output("[color=SNOW]è·³è½¬åˆ°[/color]" + str(id) + "[color=SNOW]å‰§æƒ…[/color]")
	dialog_manager.debug_jump_data(id)

func _execute_info_command():
	# æ‰§è¡ŒinfoæŒ‡ä»¤
	var info = dialog_manager.debug_get_info()
	_update_output("[color=BEIGE]æ¸¸æˆä¿¡æ¯ï¼š[/color]" + info)

func _execute_set_name_command(name: String):
	dialog_interface.set_character_name(name)
	_update_output("è®¾ç½®å¯¹è¯äººç‰©å§“åï¼š" + name)
	pass
	
func _execute_set_content_command(content: String):
	dialog_interface.set_content(content, 0.06)
	_update_output("è®¾ç½®å¯¹è¯å†…å®¹ï¼š" + content)
	pass
	
func _execute_list_dialog_datas_command():
	# æ˜¾ç¤ºç« èŠ‚IDåˆ—è¡¨
	var list = dialog_manager.debug_get_dialog_data_list()
	var list_string: String
	_update_output("å‰§æƒ…åˆ—è¡¨ï¼š \n" + ", \n".join(list))
	pass
	
func _execute_help_command():
	# æ‰§è¡ŒhelpæŒ‡ä»¤
	_update_output("æ§åˆ¶å°å¯ç”¨æŒ‡ä»¤: \n \
	help - æŸ¥çœ‹å¯ç”¨æŒ‡ä»¤å¸®åŠ©\n \
	next - å¯¹è¯ä¸‹ä¸€ä¸ª \n \
	stop - åœæ­¢å¯¹è¯ \n \
	start - å¼€å§‹å¯¹è¯ \n \
	jump_cur <id> - è·³è½¬åˆ°æŒ‡å®šå¯¹è¯å¥\n \
	jump_data <id> - è·³è½¬åˆ°æŒ‡å®šç« èŠ‚\n \
	set_name <name> - è®¾ç½®å¯¹è¯äººç‰©å§“å\n \
	set_content <content> - è®¾ç½®å¯¹è¯å†…å®¹\n \
	load_ks - æ‰“å¼€æ–‡ä»¶å¯¹è¯æ¡†åŠ è½½Konadoè„šæœ¬\n \
	info - æ‰“å°æ¸¸æˆä¿¡æ¯\n \
	list_dialog_datas - æŸ¥çœ‹å‰§æƒ…æ–‡ä»¶åˆ—è¡¨\n \
	history - å†å²æŒ‡ä»¤
	clean - æ¸…å±\n \
	clean_history - æ¸…ç©ºå†å²æŒ‡ä»¤")

func _execute_clean_command():
	# æ¸…å±æŒ‡ä»¤
	history_box.set_text("")

func _execute_history_command():
	for command in command_history:
		_update_output("[color=CYAN]history: [/color]" + command)
		
func _execute_clean_history_command():
	# æ¸…ç©ºå†å²
	command_history.clear()

func _execute_load_ks_command():
	ks_file_loader.popup()
	pass

# åŠ è½½Pluliterè„šæœ¬
func _on_ks_file_sel(path: String):
	_update_output("åŠ è½½è·¯å¾„ï¼š" + path)
	var load_data = pscripts.process_scripts_to_data(path)
	var error = await dialog_manager.debug_load_dialog_data(load_data)
	if error:
		_update_output("åŠ è½½æˆåŠŸï¼")
	elif !error:
		_update_output("åŠ è½½å¤±è´¥")
